USAGOV-2177: USAGOV-2177 breadcrumbs for directory pages

DaleMHFrey opened the pull request:
<!--- Provide a link to the Jira ticket -->
https://cm-jira.usa.gov/browse/USAGOV-2177

## Description
For State & Agency pages, relocated breadcrumb logic to the template file, and made it look up the title of the node's parents.

## Type of Changes
<!--- Put an `x` in all the boxes that apply. -->
- [ ] New Feature
- [ ] Bugfix
- [x] Frontend (Twig, Sass, JS)
- [ ] Drupal Config (requires "drush cim")
- [ ] New Modules (requires rebuild)
- [ ] Documentation
- [ ] Infrastructure
- [ ] Other

## Testing Instructions
Go to a state page like `/states/alaska` or an agency page like `/agencies/u-s-abilityone-commission` - compare what you see on dev vs Prod. As said in the ticket:
**Good:** The U.S. and its government 
**Bad:** About the U.S. and its government

## Post PR Approval Instructions

Follow these steps as soon as you merge the new changes.

1. Go to the [USAGov Circle CI project](https://app.circleci.com/pipelines/github/usagov/usagov-2021).
2. Find the commit of this pull request.
3. Build and deploy the changes.
4. Update the Jira ticket by changing the ticket status to `Review in Test` and add a comment. State whether the change is already visible on [cms-dev.usa.gov](http://cms-dev.usa.gov/) and [beta-dev.usa.gov](http://beta-dev.usa.gov/), or if the deployment is still in process.

Commit:
USAGOV-2177: USAGOV-2177 breadcrumbs for directory pages

Commit:
USAGOV-2177: lint errors

omerida commented on web/themes/custom/usagov/usagov.theme:
```diff
@@ -444,3 +444,65 @@ function usagov_theme_suggestions_alter(array &$suggestions, array $variables, $
   }
 }
+
+function usagov_preprocess_breadcrumb(&$variables) {
```
I wonder if a theme preprocess function is the right spot or if  we should use `hook_system_breadcrumb_alter()`. The homepage titles are also hard coded in the taxonomy datalayer and published pages report. If they could all get the data from one place by fetching the breadcrumb for a node, that would be ideal Does this function affect the published-pages.csv file and datalayer JSON on the page?

DaleMHFrey replied:
While I am keeping this function here, I have relocated the core of the logic to TaxonomyDatalayerBuilder. Do you want to review again to see if this looks better?

omerida commented on web/themes/custom/usagov/usagov.theme:
```diff
@@ -444,3 +444,65 @@ function usagov_theme_suggestions_alter(array &$suggestions, array $variables, $
+
+function usagov_preprocess_breadcrumb(&$variables) {
+
+  $database = \Drupal::database();
```
Let's avoid direct SQL queries, there are potential security issues plus we have to then be aware of potential schema changes in these tables when upgrading tables. This site is handy for working with Drupals entity queries: https://www.drupalatyourfingertips.com/queries

You can find similar lookups in the datalayer code and usagov menu block modules. I'll add examples below.

omerida commented on web/themes/custom/usagov/usagov.theme:
```diff
@@ -444,3 +444,65 @@ function usagov_theme_suggestions_alter(array &$suggestions, array $variables, $
+
+  $database = \Drupal::database();
+
+  if (strpos($_SERVER['REQUEST_URI'], 'agencies/') !== FALSE) {
```
to test if the current page is the federal agency page, see `TaxonomyDatalayerBuilder::isFederalDirectoryIndex()`

```php
// get current node
$node = \Drupal::routeMatch()->getParameter('node');
// check if agency index is english page
if ($this->node->toUrl()->toString() === TaxonomyDatalayerBuilder::AGENCY_INDEX_URL_EN) {
```

Similar test for Spanish page, uses a different constant.

omerida commented on web/themes/custom/usagov/usagov.theme:
```diff
@@ -444,3 +444,65 @@ function usagov_theme_suggestions_alter(array &$suggestions, array $variables, $
+  if (strpos($_SERVER['REQUEST_URI'], 'agencies/') !== FALSE) {
+    $variables['breadcrumb'] = [];
+    $variables['breadcrumb'][] = ['text' => 'Home', 'url' => 'https://www.usa.gov/'];
+    $nid = $database->query("SELECT path FROM path_alias WHERE alias LIKE '%about-the-us%'")->fetchField();
+    $nid = str_replace('/node/', '', $nid);
+    $title = $database->query("SELECT title FROM node_field_data WHERE nid = {$nid}")->fetchField();
```
Lookup the internal based on alias, this should return '/node/NID`. Extract nid from there, then use `Node::load($nid)` to get the Node entity and get the title from it.

```
$path = \Drupal::service('path_alias.manager')->getPathByAlias($url);

DaleMHFrey replied:
I have relocated and refactored the core of the logic to TaxonomyDatalayerBuilder. Do you want to review again to see if this looks better?

omerida commented on web/themes/custom/usagov/usagov.theme:
```diff
@@ -444,3 +444,65 @@ function usagov_theme_suggestions_alter(array &$suggestions, array $variables, $
+    $nid = str_replace('/node/', '', $nid);
+    $title = $database->query("SELECT title FROM node_field_data WHERE nid = {$nid}")->fetchField();
+    $variables['breadcrumb'][] = ['text' => $title, 'url' => '/about-the-us'];
+    $nid = $database->query("SELECT path FROM path_alias WHERE alias LIKE '%agency-index%'")->fetchField();
```
same as above but for Spanish, you might have to include the langcode as the second parameter to `getPathByAlias()`

omerida commented on web/themes/custom/usagov/usagov.theme:
```diff
@@ -444,3 +444,65 @@ function usagov_theme_suggestions_alter(array &$suggestions, array $variables, $
+  if (strpos($_SERVER['REQUEST_URI'], 'states/') !== FALSE) {
+    $variables['breadcrumb'] = [];
+    $variables['breadcrumb'][] = ['text' => 'Home', 'url' => 'https://www.usa.gov/'];
+    $nid = $database->query("SELECT path FROM path_alias WHERE alias LIKE '%about-the-us%'")->fetchField();
```
use path alias manager to get the node and then its title

omerida commented on web/themes/custom/usagov/usagov.theme:
```diff
@@ -444,3 +444,65 @@ function usagov_theme_suggestions_alter(array &$suggestions, array $variables, $
+    $nid = str_replace('/node/', '', $nid);
+    $title = $database->query("SELECT title FROM node_field_data WHERE nid = {$nid}")->fetchField();
+    $variables['breadcrumb'][] = ['text' => $title, 'url' => '/about-the-us'];
+    $nid = $database->query("SELECT path FROM path_alias WHERE alias LIKE '%state-governments%'")->fetchField();
```
use path alias manager to get the node and then its title

DaleMHFrey replied:
I have relocated and refactored the core of the logic to TaxonomyDatalayerBuilder. Do you want to review again to see if this looks better?

omerida commented on web/themes/custom/usagov/usagov.theme:
```diff
@@ -444,3 +444,65 @@ function usagov_theme_suggestions_alter(array &$suggestions, array $variables, $
+  if (strpos($_SERVER['REQUEST_URI'], 'agencias/') !== FALSE) {
+    $variables['breadcrumb'] = [];
+    $variables['breadcrumb'][] = ['text' => 'Home', 'url' => 'https://www.usa.gov/'];
+    $nid = $database->query("SELECT path FROM path_alias WHERE alias LIKE '%acerca-de-estados-unidos%'")->fetchField();
```
use path alias manager to get the node and then its title

omerida commented on web/themes/custom/usagov/usagov.theme:
```diff
@@ -444,3 +444,65 @@ function usagov_theme_suggestions_alter(array &$suggestions, array $variables, $
+    $variables['breadcrumb'][] = ['text' => Drupal::service('title_resolver')->getTitle(\Drupal::request(), \Drupal::routeMatch()->getRouteObject())];
+  }
+
+  if (strpos($_SERVER['REQUEST_URI'], 'estados/') !== FALSE) {
```
same as above, get the node for the current match request to test the full URL path. This will match any URL that has `estados/` in it

omerida commented on web/themes/custom/usagov/usagov.theme:
```diff
@@ -444,3 +444,65 @@ function usagov_theme_suggestions_alter(array &$suggestions, array $variables, $
+  if (strpos($_SERVER['REQUEST_URI'], 'estados/') !== FALSE) {
+    $variables['breadcrumb'] = [];
+    $variables['breadcrumb'][] = ['text' => 'Home', 'url' => 'https://www.usa.gov/'];
+    $nid = $database->query("SELECT path FROM path_alias WHERE alias LIKE '%acerca-de-estados-unidos%'")->fetchField();
```
use path alias manager to get the node and then its title

omerida commented on web/themes/custom/usagov/usagov.theme:
```diff
@@ -444,3 +444,65 @@ function usagov_theme_suggestions_alter(array &$suggestions, array $variables, $
+    $nid = str_replace('/node/', '', $nid);
+    $title = $database->query("SELECT title FROM node_field_data WHERE nid = {$nid}")->fetchField();
+    $variables['breadcrumb'][] = ['text' => $title, 'url' => '/acerca-de-estados-unidos'];
+    $nid = $database->query("SELECT path FROM path_alias WHERE alias LIKE '%gobiernos-estatales%'")->fetchField();
```
use path alias manager to get the node and then its title

whitealbum committed:
Merge branch 'dev' into USAGOV-2177

Commit:
USAGOV-2177: PR review

Commit:
USAGOV-2177: PR review

Commit:
USAGOV-2177: PR review

Commit:
USAGOV-2177: PR review

omerida commented:
Looks like the changes should be fine, will test locally but phpcs is flagging issues: <https://app.circleci.com/pipelines/github/usagov/usagov-2021/12563/workflows/9f74e44b-f113-4b84-bd80-71c6101f099d/jobs/31552/parallel-runs/0/steps/0-104> @DaleMHFrey

omerida commented on web/modules/custom/usa_twig_vars/src/TaxonomyDatalayerBuilder.php:
```diff
@@ -57,6 +45,90 @@ public function __construct(
     $this->isFront = $isFront ? 'homepage' : 'not_homepage';
   }
 
+  public static function ABOUT_GOVT_EN() {
```
codesniff is complaining about the method name. It wants camelcase, something like 'getAboutGovtEn()` would work. This goes for all the static method names.

omerida commented on web/themes/custom/usagov/templates/breadcrumb.html.twig:
```diff
@@ -7,40 +7,6 @@
-        { 'url': '/agency-index', 'text': drupal_field('title', 'node', 629) },
-        { 'text': drupal_title() }
-    ] %}
-{% elseif path('<current>') matches '#\/states\/*#' %}
-    {% set breadcrumb = [
-        { 'url': "/", 'text': 'Home' },
-        { 'url': "/about-the-us", 'text': 'About the U.S. and its government' },
-        { 'url': "/state-governments", 'text': 'State governments' },
-        { 'text': drupal_title() }
-    ] %}
```
Not sure what the cause is. I visit <https://localhost/states/wyoming>, and the breadcrumb just shows the homepage
![image](https://github.com/user-attachments/assets/bb46f1b6-119a-40dc-8be9-c94aa08c3770)

omerida replied:
Likewise on agency pages like <https://localhost/agencies/u-s-capitol-police>

![image](https://github.com/user-attachments/assets/8ebac488-71ba-4daf-9a7a-096e52606b77)

omerida replied:
also on Spanish pages https://localhost/es/agencias/administracion-de-alimentos-y-medicamentos and https://localhost/es/estados/alaska

The taxonomy datalayer stuff is correct... I think this template is still needed since breadcrumb and the datalayer aren't coupled.

omerida replied:
Not sure if my set up is borked. I did a  `drush cr` and `usagov_preprocess_breadcrumb()` doesn't get called.

omerida replied:
I had to restart my docker container and now the theme hook is getting called and building the breadcrumbs.

omerida commented on web/modules/custom/usa_twig_vars/src/TaxonomyDatalayerBuilder.php:
```diff
@@ -57,6 +45,90 @@ public function __construct(
+    if ($ret === false) {
+      $sysPath = \Drupal::service('path_alias.manager')->getPathByAlias(self::ABOUT_URL_ES());
+      $nid = str_replace('/node/', '', $sysPath);
+      $ret = Node::load($nid)->getTitle();
```
the published-pages.csv is failing at this line for me. 

`drush usapubcsv '--uri=http://localhost/' modules/custom/usagov_ssg_postprocessing/files/published-pages.csv`

![image](https://github.com/user-attachments/assets/81745812-124e-4e6d-8d4f-2721ae25225e)

DaleMHFrey replied:
After the refactoring I am no longer able to reproduce this problem on my local - do you want to try again?

omerida replied:
sure, I'll try again.

omerida commented on web/themes/custom/usagov/usagov.theme:
```diff
@@ -444,3 +445,42 @@ function usagov_theme_suggestions_alter(array &$suggestions, array $variables, $
+
+  if (strpos($node->toUrl()->toString(), 'agencies/') !== false) {
+    $variables['breadcrumb'] = [];
+    $variables['breadcrumb'][] = ['text' => 'Home', 'url' => 'https://www.usa.gov/'];
```
Do we have to hard-code the prod domain here. Can we use `'url' => '/'` ?

omerida commented on web/themes/custom/usagov/usagov.theme:
```diff
@@ -444,3 +445,42 @@ function usagov_theme_suggestions_alter(array &$suggestions, array $variables, $
+  $node = \Drupal::routeMatch()->getParameter('node');
+  if (empty($node)) return;
+
+  if (strpos($node->toUrl()->toString(), 'agencies/') !== false) {
```
```suggestion
  if (str_starts_with('/agencies/', $node->toUrl()->toString())) {
```

omerida commented on web/themes/custom/usagov/usagov.theme:
```diff
@@ -444,3 +445,42 @@ function usagov_theme_suggestions_alter(array &$suggestions, array $variables, $
+    $variables['breadcrumb'][] = ['text' => Drupal::service('title_resolver')->getTitle(\Drupal::request(), \Drupal::routeMatch()->getRouteObject())];
+  }
+
+  if (strpos($node->toUrl()->toString(), 'states/') !== false) {
```
```suggestion
  if (str_starts_with('/states/', $node->toUrl()->toString())) {
```

omerida commented on web/themes/custom/usagov/usagov.theme:
```diff
@@ -444,3 +445,42 @@ function usagov_theme_suggestions_alter(array &$suggestions, array $variables, $
+    $variables['breadcrumb'][] = ['text' => Drupal::service('title_resolver')->getTitle(\Drupal::request(), \Drupal::routeMatch()->getRouteObject())];
+  }
+
+  if (strpos($node->toUrl()->toString(), 'agencias/') !== false) {
```
```suggestion
  if (str_starts_with('/es/agencias/', $node->toUrl()->toString())) {
```

omerida commented on web/themes/custom/usagov/usagov.theme:
```diff
@@ -444,3 +445,42 @@ function usagov_theme_suggestions_alter(array &$suggestions, array $variables, $
+    $variables['breadcrumb'][] = ['text' => Drupal::service('title_resolver')->getTitle(\Drupal::request(), \Drupal::routeMatch()->getRouteObject())];
+  }
+
+  if (strpos($node->toUrl()->toString(), 'estados/') !== false) {
```
```suggestion
  if (str_starts_with('/es/estados/', $node->toUrl()->toString())) {
```

whitealbum committed:
Merge branch 'dev' into USAGOV-2177

Commit:
USAGOV-2177: lint

Commit:
USAGOV-2177: lint

Commit:
USAGOV-2177: lint

Commit:
USAGOV-2177: lint

omerida committed:
Merge branch 'dev' into USAGOV-2177

omerida commented on web/themes/custom/usagov/usagov.theme:
```diff
@@ -465,3 +466,68 @@ function usagov_theme_suggestions_alter(array &$suggestions, array $variables, s
+    return;
+  }
+
+  if (str_starts_with('/agencies/', $node->toUrl()->toString())) {
```
Sorry but I got needle and haystack backwards on all of these, it should be the `$node->` part first

```php
if ($node->toUrl()->toString(), str_starts_with('/agencies/')) {
```

omerida commented on web/themes/custom/usagov/usagov.theme:
```diff
@@ -465,3 +466,68 @@ function usagov_theme_suggestions_alter(array &$suggestions, array $variables, s
+    $variables['breadcrumb'][] = ['text' => Drupal::service('title_resolver')->getTitle(\Drupal::request(), \Drupal::routeMatch()->getRouteObject())];
+  }
+
+  if (str_starts_with('/es/agencias/', $node->toUrl()->toString())) {
```
flip haystack and needle

omerida commented on web/themes/custom/usagov/usagov.theme:
```diff
@@ -465,3 +466,68 @@ function usagov_theme_suggestions_alter(array &$suggestions, array $variables, s
+    $variables['breadcrumb'][] = ['text' => Drupal::service('title_resolver')->getTitle(\Drupal::request(), \Drupal::routeMatch()->getRouteObject())];
+  }
+
+  if (str_starts_with('/es/estados/', $node->toUrl()->toString())) {
```
flip haystack and needle

Commit:
USAGOV-2177: pr-review

akf committed:
Merge branch 'dev' into USAGOV-2177

omerida committed:
Merge branch 'dev' into USAGOV-2177

omerida committed:
USAGOV-2177: Fixes looking up Spanish path aliases.
Fixes phpstan errors.
Replace more hardcoded titles and paths with methods for dynamic lookups.
Removes unused breadcrumb template suggestion.
Restores wizard breadcrumbs.

omerida committed:
Merge branch 'dev' into USAGOV-2177
